lib/LTO/LTO.cpp

```diff
@@ -0,0 +1,177 @@
+//===--- LTO.cpp - Swift LTO ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/LTO/LTO.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/IRGenOptions.h"
+#include "swift/AST/IRGenRequests.h"
+#include "swift/ClangImporter/ClangImporter.h"
+#include "swift/ClangImporter/ClangImporterOptions.h"
+#include "swift/Frontend/InputFile.h"
+#include "swift/Frontend/ModuleInterfaceLoader.h"
+#include "swift/Frontend/PrintingDiagnosticConsumer.h"
+#include "swift/SIL/SILModule.h"
+#include "swift/Serialization/SerializedModuleLoader.h"
+#include "swift/Serialization/SerializedSILLoader.h"
+#include "swift/Serialization/Validation.h"
+#include "swift/Subsystems.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Bitcode/BitcodeWriter.h"
+#include "llvm/IR/Module.h"
+
+namespace swift {
+
+namespace lto {
+
+using namespace llvm;
+
+class InputModule {
+public:
+  explicit InputModule(Identifier ModuleName) : ModuleName(ModuleName) {}
+  Identifier ModuleName;
+};
+
+class LTOPipelineImpl {
+  SmallVector<StringRef, 4> RuntimeLibraryPaths;
+  SmallVector<StringRef, 4> RuntimeLibraryImportPaths;
+  StringRef RuntimeResourcePath;
+  SmallVector<InputModule, 2> Modules;
+  LangOptions LangOpts;
+  ClangImporterOptions ClangOpts;
+  TypeCheckerOptions TCOpts;
+  SearchPathOptions SearchPathOpts;
+  SourceManager SM;
+  DiagnosticEngine Diags;
+  PrintingDiagnosticConsumer PrintDiags;
+  std::unique_ptr<ASTContext> Ctx;
+  MemoryBufferSerializedModuleLoader *MBL;
+
+public:
+  LTOPipelineImpl(SmallVector<StringRef, 4> RuntimeLibraryPaths,
+                  SmallVector<StringRef, 4> RuntimeLibraryImportPaths,
+                  StringRef RuntimeResourcePath)
+      : RuntimeLibraryPaths(RuntimeLibraryPaths),
+        RuntimeLibraryImportPaths(RuntimeLibraryImportPaths),
+        RuntimeResourcePath(RuntimeResourcePath), Diags(SM) {}
+
+  bool addModule(std::unique_ptr<MemoryBuffer> Buffer) {
+    serialization::ExtendedValidationInfo extendedInfo;
+    serialization::ValidationInfo info = serialization::validateSerializedAST(
+        Buffer->getBuffer(), &extendedInfo);
+    if (info.status != serialization::Status::Valid) {
+      return true;
+    }
+
+    if (!Ctx) {
+      Ctx.reset(createASTContext(info, extendedInfo));
+    }
+
+    MBL->registerMemoryBuffer(info.name, std::move(Buffer));
+
+    Modules.emplace_back(Ctx->getIdentifier(info.name));
+    return false;
+  }
+
+  bool emitLLVMModules(ModuleOutStreamFn ModuleOutStream) {
```

Comment:
`GetStream` seems like a nicer name given that the function gets the stream to write to.